src/functional/fencil_processors/builders/cpp/build.py

```diff
@@ -135,61 +136,82 @@ def _get_python_module_suffix():
     return importlib.machinery.EXTENSION_SUFFIXES[0][1:]
 
 
-class CMakeProject:
-    """Work with CMake projects on the file system."""
-
-    _folder: Optional[pathlib.Path] = None
-    _name: str
-    _extension: str
-    _cmakelists: str
-    _sources: Dict[str, str]
-
-    def __init__(
-        self,
-        name: str,
-        dependencies: Sequence[source_modules.LibraryDependency],
-        sources: Dict[str, str],
-    ):
-        self._name = name
-        self._extension = _get_python_module_suffix()
-        self._cmakelists = _render_cmakelists(name, dependencies, list(sources.keys()))
-        self._sources = sources
-
-    @classmethod
-    def get_binary_path(
-        cls, root_folder: pathlib.Path, name: str, extension: str = _get_python_module_suffix()
-    ) -> pathlib.Path:
-        return root_folder / _BUILD_SUBDIR / "bin" / (name + "." + extension)
+@dataclass(frozen=True)
```

Comment:
The thought behind making this a frozen dataclass is to make it hashable which could be beneficial for caching. It might be considered premature and comes at the cost that some of the properties (see below) have to be dynamically recomputed multiple times (although they are cheap).

There are two ways of solving this: either CMakeProject itself is again made stateful and the properties can be computed at construction (hashability is gone then) or the class is split into a dataclass which is hashable and a non-hashable class which can be created from it (this second class could be free of any hashing concerns and know only how to deal with files and the `cmake` command).

I am not sure it is worth deciding either way before we start implementing both the caching and the second build system. Both might influence the decision.

Reply:
I would like to do that in the following PR, where I am touching the `CMakeProject` anyway. Also to avoid conflicts.

Reply:
If you think this form is better for this PR and you remove it in the next, it's fine for me.

Reply:
Rico, can you put a TODO on this before merging?